### ISSUE TITLE
Update 2.5.0 to 2.5.3: `fatal: bad object refs/remotes/origin/HEAD`

### ISSUE THREAD

**Issue by koppor** [git-for-windows/git#423]
After updating from 2.5.0 to 2.5.3, I get the error message `fatal: bad object refs/remotes/origin/HEAD` when doing `git gc`. `git fsck` outputs

    error: refs/remotes/origin/HEAD: invalid sha1 pointer 0000000000000000000000000000000000000000

Content of `.git/refs/remotes/origin/HEAD` is `ref: refs/remotes/origin/master`. The file `.git/refs/remotes/origin/master` doesn't exist. The branch `master` at `origin` also doesn't exist. It is the branch `template`.

**Solution** Edit `.git/refs/remotes/origin/HEAD` manually and let it contain `ref: refs/remotes/origin/template`.

That solution is only a quick hack and it should be investigated, why git 2.5.3 deletes `refs/remotes/origin/template` and replaces it by `refs/remotes/origin/HEAD` pointing to an invalid branch.

<s>I'm currently investigating possible solutions such as http://stackoverflow.com/q/25797315/873282.</s>

**Comment by Radrik5** [comment]
What does `git reflog origin/HEAD` return?

**Comment by koppor** [comment]
warning: ignoring dangling symref refs/remotes/origin/HEAD.
    warning: ignoring dangling symref refs/remotes/origin/HEAD.
    fatal: ambiguous argument 'origin/HEAD': unknown revision or path not in the working tree.
    Use '--' to separate paths from revisions, like this:
    'git <command> [<revision>...] -- [<file>...]'

I am not sure how to reproduce the issue as it only happens if the upstream repository does not contain a master branch and if it was cloned with git 1.9.x, then used with git 2.5.0, then with git 2.5.2. Feel free to close this issue. Google will surely find it if someone else discovers the same issue.

**Comment by dscho** [comment]
I guess the real solution is to let `git gc` remove a bogus left-over `refs/remotes/origin/HEAD`. But that would probably be a discussion for the [Git mailing list](mailto:git@vger.kernel.org), as it is not limited to Windows.

**Comment by Radrik5** [comment]
I guess the real solution is not to create a bogus `refs/remotes/origin/HEAD`. It should point at `template` if remote `HEAD` points at `template` or it shouldn't exist if remote `HEAD` doesn't exist.

**Comment by dscho** [comment]
@Radrik5 maybe you point into the direction of a possible MCVE: if you clone from a repository whose active branch is `template`, then the active branch in the remote repository is switched to `master` and `template` is removed, then fetch (with --prune), we should be able to replicate the error, generating the MCVE that every good bug report has. Don't you think?

**Comment by koppor** [comment]
In my case, I used https://github.com/latextemplates/uni-stuttgart-dissertation-template, but there, the remote repository did not switch to `master` at all. On Thursday, I'll have access to a machine with git 2.5.0 and with the repository checked out (using git 1.x). Then, I'll update to git 2.5.3 and report back, what happened.

**Comment by dscho** [comment]
And indeed, I could produce an MCVE: https://github.com/dscho/git/commit/ec2bd787adbe75d4fba01f4977e0b24cc4ef48a6

**Comment by koppor** [comment]
I checked with the other machine, everything worked fine there. @dscho Yes, your MCVE indeed resembles the steps I did.

I wonder, why this came up when switchting from 1.9.x to 2.5.x. I could normally push and pull with old versions of git without any issues. I think, I never called `git gc` though...

Following is the directory listing of `.git/refs/remotes/origin`:

```
11.03.2014  16:04                32 HEAD
03.06.2015  10:44                41 lncs_as
29.05.2015  20:45                41 template
```

Is `HEAD` not used in the 1.9.x branch, but in the 2.5.x branch?
